### PR TITLE
fix(git): simple swap to ioutil.Discard for progress

### PR DESCRIPTION
**Pull request by bkono**
@thbishop-intuit for #149 were you literally looking to silence the output like this? Or did you want something more complete to flow through, like os.Stdout when verbose or debug logging is set, and discarded the rest of the time? Your pointers were dead on to where the output was coming from btw.

### CODE REVIEW

**Comment by bkono** [comment]
Since the coverage on the git package was so low, I ran into the change not being tested. Took that as a sign to do something slightly smarter, and set the Writer up as a functional option to pass in. Also updated the service/main to have the LogLevel control whether it is os.Stdout or ioutil.Discard, giving us a path to actually see the git progress output during dev and testing when needed. Open to thoughts on it, @thbishop-intuit.

File: service/internal/git/gitClient.go [hunk 1]
```diff
@@ -51,43 +53,60 @@ func (g gitSvcImpl) Checkout(w *git.Worktree, opts *git.CheckoutOptions) error {
 	return w.Checkout(opts)
 }
 
+// Option is a function for configuring the BasicClient
+type Option func(*BasicClient)
+
+func WithProgressWriter(w io.Writer) Option {
```
Comment (bkono) [inline service/internal/git/gitClient.go:59]: Simple functional opts pattern to allow whatever writer we want to come in.

File: service/internal/git/gitClient.go [hunk 2]
```diff
@@ -51,43 +53,60 @@ func (g gitSvcImpl) Checkout(w *git.Worktree, opts *git.CheckoutOptions) error {
...
+func NewHTTPSBasicClient(user, pass string, opts ...Option) (BasicClient, error) {
+	auth := &http.BasicAuth{
+		Username: user,
+		Password: pass,
+	}
+
+	return newBasicClient(auth, opts...), nil
+}
+
+func newBasicClient(auth transport.AuthMethod, opts ...Option) BasicClient {
```
Comment (bkono) [inline service/internal/git/gitClient.go:95]: Refactored away the commonality, since the other two were just wrappers changing the AuthMethod.

File: service/internal/git/gitClient.go [hunk 3]
```diff
@@ -51,43 +53,60 @@ func (g gitSvcImpl) Checkout(w *git.Worktree, opts *git.CheckoutOptions) error {
...
+	cl := BasicClient{
+		auth:    auth,
 		mu:      &sync.Mutex{},
 		git:     gitSvcImpl{},
 		fs:      os.DirFS(os.TempDir()),
 		baseDir: os.TempDir(),
-	}, nil
+		pw:      ioutil.Discard,
```
Comment (bkono) [inline service/internal/git/gitClient.go:102]: Defaulting to discard unless overridden, which will happen currently in `service/main` when env.LogLevel is debug.

File: service/internal/git/gitClient_test.go [hunk 4]
```diff
@@ -10,52 +11,71 @@ import (
...
+func (g *mockGitSvc) PlainClone(path string, isBare bool, o *git.CloneOptions) (*git.Repository, error) {
+	g.cloneOpts = o
 	return nil, nil
 }
 
-func (g mockGitSvc) PlainOpen(path string) (*git.Repository, error) {
+func (g *mockGitSvc) PlainOpen(path string) (*git.Repository, error) {
+	g.plainOpened = true
+	if strings.Contains(path, "Err") {
+		return nil, errors.New("plain open goes boom")
```
Comment (bkono) [inline service/internal/git/gitClient_test.go:28]: Will do a separate PR to fix test coverage on the git package. This is just ground work, and opening a path to get coverage on the lines I wanted changed.

File: service/main.go [hunk 5]
```diff
@@ -99,3 +84,28 @@ func setLogLevel(logger *log.Logger, logLevel string) {
...
+func gitClient(env env.Vars, logger log.Logger) git.BasicClient {
+	var cl git.BasicClient
+	var err error
+
+	var opts []git.Option
+	if env.LogLevel == "DEBUG" {
+		opts = append(opts, git.WithProgressWriter(os.Stdout))
+	}
+
+	if env.GitAuthMethod == "https" {
```
Comment (bkono) [inline service/main.go:97]: Cringed when I brought over the if / else, but the refactor to remove that seemed like unnecessary bloat on the Progress Writer change.

**Comment by bkono** [comment]
So much for avoiding unnecessary bloat 😆 . Refactored the git tests to bump coverage >80% with the changes.

**Review by pclata — APPROVED** [review]
Nice work!
